spec/services/badge_granter_spec.rb

```diff
@@ -482,14 +482,17 @@
       expect(UserBadge.find_by(user_id: user.id, badge_id: Badge::GreatPost)).to eq(nil)
     end
 
-    it 'calls user_badge_granted on backfilled badge' do
+    it "calls 'user_badge_granted' DiscourseEvent on backfilled badge" do
       post = create_post(user: user)
       action = PostActionCreator.like(liker, post).post_action
 
-      Badge.any_instance.expects(:trigger_badge_granted_event).with(user.id).once
-      Badge.any_instance.expects(:trigger_badge_granted_event).with(liker.id).once
+      events = DiscourseEvent.track_events(:user_badge_granted) do
+        BadgeGranter.process_queue!
+      end
 
-      BadgeGranter.process_queue!
+      expect(events.length).to eq(2)
+      expect(events[0][:params]).to eq([Badge::FirstLike, liker.id])
+      expect(events[1][:params]).to eq([Badge::Welcome, user.id])
```

Comment:
@jjaffeux This is what I meant by testing for the expected outcome instead of the implementation detail.

Reply:
I had understood, and it was how my test was in the first place if you look at commits history.

Reply:
Yup so I was wondering why it was changed to a mock in a follow up commit.